### PR TITLE
Implement isend/irecv functions.

### PR DESCRIPTION
**Pull request by bangerth**
This is a work-in-progress thing. For a long time already, I wanted to implement MPI's isend/irecv functions in such a way that (i) they can send arbitrary objects via our packing/unpacking functionality, and (ii) so that they return `std::future` objects. If one were to do this in a reasonable way, it would reduce the amount of code one needs by 20 or 30 lines for each call. Moreover, I can then implement what I'm really interested in: an `irecv_and_then()` function that, once it has obtained the message, unpacks it and runs a user-defined operation on it. This is really what one needs: get the data and do something with it, in one place, and not scattered across a lot of code where one posts the receive operation somewhere and somewhere else waits for it and does something with the data.

To do this in a useful way, one has to hide waiting for the communication to finish on a separate thread. This patch provides a trial case for this, where I'm still implementing everything in a simple testcase to make it easier to test a few things without having to recompile everything all the time.

**I need some help:** The way I'm implementing the send operation calls `MPI_Isend` on the calling thread, and puts an `MPI_Wait` on a separate thread that eventually makes the `std::future` ready. For the receive operation, I'm running an `MPI_Probe` on a separate thread followed by the `MPI_Recv`. In either case, a wait-type operation is run on a separate thread. This is only useful if MPI implementations do this without a spin-lock since otherwise one ends up with a separate thread for each posted `irecv` that has a spin-lock going.

What I could use some help with is for others to check what they see with their implementations. The testcase in this PR sends ten messages to each of the other processes, with the receives posted previously. I'm imposing a ten-second delay between posting receives and sends during which there are `10 * (n_mpi_processes - 1)` threads waiting in an `MPI_Probe` call; each process' main thread is also waiting for a `std::future`. Can you run the test here (which will execute with 2 and 4 processes) and, in a separate window, run `top` to see what is happening? For example by doing
```
  ctest -R irecv.*debug
  cd tests/mpi/isend_irecv_01.debug/
  time mpirun -np 2 ./isend_irecv_01.debug
  time mpirun -np 4 ./isend_irecv_01.debug
```
On my system, the `isend_irecv_01.debug` process does not even show up in `top`, i.e., apparently an efficient mutex and not a spin-lock is used. Is this true with other MPI implementations as well? Here is my `time` output, showing that the program really does appear to not take any time while waiting for messages to come in:
```
> time mpirun -np 2 ./isend_irecv_01.debug 

real    0m11.675s
user    0m0.269s
sys     0m0.187s


> time mpirun -np 4 ./isend_irecv_01.debug 

real    0m11.660s
user    0m0.497s
sys     0m0.235s
```

/rebuild

### CODE REVIEW

**Comment by tamiko** [comment]
... but what works is simply spawning a thread via `std::async` and doing a standard synchronous exchange on that thread. Maybe it is of interest for you (code simplified):

```
class SynchronizationDispatch
{
public:
  SynchronizationDispatch(const std::function<void()> &async_payload)
      : async_payload_(async_payload)
  {
  }

  ~SynchronizationDispatch()
  {
    /* Executes in serial, non thread-parallel context: */

    if (payload_status_.valid()) {
      payload_status_.wait();
    } else {
      async_payload_();
    }
  }

  DEAL_II_ALWAYS_INLINE inline void check(...)
  {
    if (...)
      payload_status_ = std::async(std::launch::async, async_payload_);
  }

private:
  const std::function<void()> async_payload_;
  std::future<void> payload_status_;
  std::atomic_int n_threads_ready_;
};
```

**Comment by tamiko** [comment]
Correct, I rely on the current implementation of `std::async` with libstdc++. I'll have to change that at some point.

I haven't considered changing the thread level to `MPI_THREAD_MULTIPLE` because of some scary stories of performance degradation - but I think that this is something that we should verify independently and benchmark quickly (ping @kronbichler).

On another note, the ucp failure with `MPI_THREAD_SERIAL` is actually a good sign in the sense that OpenMPI does seem to implement a thread-local message queue - in this sense we'll have to see how much the forced synchronization of `MPI_THREAD_MULTIPLE` will degrade performance. Judging from a quick literature search there seem to be quite some contention in the message queue in OpenMPI with threads anyway, so `MPI_THREAD_MULTIPLE` might not be much of a performance hit after all... (let me find that reference for you)

**Comment by tamiko** [comment]
Regarding the current logic in `mpi.cc`:
```
int wanted = MPI_THREAD_SERIALIZED;
ierr       = MPI_Init_thread(&argc, &argv, wanted, &provided);
AssertThrowMPI(ierr);

// disable for now because at least some implementations always return MPI_THREAD_SINGLE.
Assert(max_num_threads==1 || provided != MPI_THREAD_SINGLE, ..);
```

What about we switch this to a more flexible logic:
```
int wanted = MPI_THREAD_MULTIPLE;
ierr       = MPI_Init_thread(&argc, &argv, wanted, &provided);
AssertThrowMPI(ierr);

// something like...
dealii_have_fully_multithreaded_mpi = (provided == MPI_THREAD_MULTIPLE)
```
and we then guard the asynchronous, multithreaded isend/irecv with `dealii_have_fully_multithreaded_mpi()` and fall back to synchronous, blocking and non-threaded otherwise. That way we don't have to worry about what MPI implementations actually support.

**Comment by kronbichler** [comment]
I have doubts whether this setup is worth it, compared to the performance penalty we might need to pay. Let us discuss a few questions I have at this point, so I can understand better:
- This code creates two new threads for each message an MPI rank sends and receives. Let us assume a typical 3D mesh of a cube with enough MPI processes, where each rank has 26 neighbors for a typical ghost exchange. How much does it cost to create 52 threads in terms of latency and compute time? I realize that good threads implementations re-use the threads from a pool, so my question is about the latency cost of that pool retrieval and possible locks used in there.
- MPI messages via MPI_Isend/Irecv are highly tuned towards low latency. One can expect a message to take around 3 to 6 microseconds (3e-6) between send and receive when everything is ready. This comes from the MPI point-to-point latency of somewhere around 0.5 microseconds plus some overhead. A full ghost exchange to all 26 neighbors takes around 20 microseconds. When running a multigrid solver, I would expect 90% of the messages to have waiting times on the order of 10 microseconds or less. The question is whether the memory allocations and deallocations, thread creation and destruction, and all other operations would be still small against this 20 microsecond target. Can we design a benchmark to try this out?
- Coming to the hardware point of view, we can distinguish between two cases. In the first case we have a fully allocated node with no hyperthreading (simultaneous multithreading), the machine gets oversubscribed and we need to rely on the operating system to slice the work of different threads onto the cores. What is the timings for these slots? If I remember correctly, these are somewhere between 1e-4 and 1e-3 seconds. The questions are if we run into the timing of these ticks, or not. In the second case we have idle threads, often hyperthreads, that can absorb the operations pretty well. The question is then whether the operating system or thread pinning schedules them on the same physical core as the main thread (in which case the hardware latency is way below 1 microsecond) or if it gets somewhere else (in that case, we likely spend more than a microsecond, because that thread might have been in a low-power/downclocked CPU core).
- In a benchmark mentioned by @tamiko above, I had to disable any attempt to have threads running the MPI communication in parallel to some computations for some MPI implementations (otherwise we would lose 2x performance), and even for the systems with the best MPI implementations it turned out that using plain MPI is statistically never slower.

If my doubts turn out true, I would be strongly opposed to applying these in the performance-critical code like https://github.com/dealii/dealii/blob/c1307078342e02435df476fc1b0588c8d482b926/include/deal.II/base/partitioner.templates.h#L91-L174 and similar places used by `LinearAlgebra::distributed::Vector` and the `MatrixFree::loop` facilities. That would mean that we have some parts of the library that use these new facilities with `promise/future`, and others that keep using plain `MPI_Isend/Irecv`.

As a side note, against the necessities in the `LinearAlgebra::distributed::Vector` class, I have two additional issues we would need to address:
- Doing an `MPI_Probe` followed by an `MPI_Recv` did show worse performance than `MPI_Irecv` on some tests I did 10 years ago, especially near the scaling limit. `MPI_Probe` is unnecessary in those places because we have the right amount of memory allocated before and have full knowledge of the communication pattern of all owners. We would need to run benchmarks and verify that this is not problematic.
- Doing `Utilities::pack/unpack` is not appropriate for the use cases mentioned above, at least not without additional work. The reason is that create another copy of data, and the use cases in the partitioner or ghost exchange are memory-bandwidth limited within the cores for large sizes, so this directly eats into the observed performance (this is documented in the 2019 matrix-free paper, Figure 13). That could be fixed by doing some `reinterpret_cast` for a data structure that might or might not create a deep copy of an array.

To summarize, it could be that all of the issues I raise above can be satisfactorily addressed. Doing this in a high-performance-compatible way will take a major effort both in benchmarking and developing the right solutions, so the question is whether a fully fledged solution is realistic. We would need volunteers for this, I cannot promise to have time on this subject apart from verifying performance at some point.

**Comment by kronbichler** [comment]
> In either case, a wait-type operation is run on a separate thread. This is only useful if MPI implementations do this without a spin-lock since otherwise one ends up with a separate thread for each posted irecv that has a spin-lock going.

@bangerth I do not really understand this comment, at least against the benchmark code you posted. If I understand the code correctly, the communication is not actually delayed by 10 seconds, and I would expect it to indeed finish quickly. This is because you post the full send and receive commands immediately, so the separate threads will finish the work quickly. Then, it is just the empty future that waits, but without being inside the `MPI_Wait` function any more. In other words, the reason we see no CPU load is that because we indeed are in the `sleep` function with the main thread in at the end of the thread's work otherwise.

I will post a remark where I would put the `sleep` to stress things.

**Comment by kronbichler** [comment]
For my last two comments, note that I might completely misunderstand how `std::future` and `std::promise` work and when precisely the additional threads are scheduled and can run in the background. (I tried to read up on the topic, but I find the documentation pretty hard to understand.)

**Comment by bangerth** [comment]
On 6/24/22 03:22, Martin Kronbichler wrote:
> @bangerth <https://github.com/bangerth> I do not really understand this 
> comment, at least against the benchmark code you posted. If I understand
> the code correctly, the communication is not actually delayed by 10
> seconds, and I would expect it to indeed finish quickly. This is because
> you post the full send and receive commands immediately, so the separate
> threads will finish the work quickly. Then, it is just the empty future
> that waits, but without being inside the |MPI_Wait| function any more. In
> other words, the reason we see no CPU load is that because we indeed are in
> the |sleep| function with the main thread in at the end of the thread's
> work otherwise.

Oh, doh! I put the sleep into the wrong place! (PR updated.)

I think that also kills the idea:
```
 > time mpirun -np 4 ./isend_irecv_01.debug

real    0m11.673s
user    0m40.570s
sys     0m0.211s
```
In other words, my MPI implementation does a busy wait for these receives, 
rather than use threads that are woken up when the data is there. That's not 
good, I don't think we should pursue this direction any further.

I hear what you say about performance. My interest wasn't so much replacing 
the existing places than experimenting with some higher level interfaces that 
make it easier to read code, in particular using irecv_and_then(). It could be 
used in places such as DoFHandler::distribute_dofs() or FETools::extrapolate() 
-- places that are not called in each inner iteration of a linear solver, for 
example. But given the busy lock above, I don't think that we should continue 
in this direction.

**Comment by bangerth** [comment]
It turns out that one doesn't actually need a separate thread. I just needed something "like" `std::future` that has a `wait()` and a `get()` function. The implementation of `isend()` and `irecv()` now set up the communication and then create such a `Utilities::MPI::Future` object and tells it what needs to be done for `wait()` and `get()` -- provided through some lambda functions.

This implementation basically encapsulates what one does by hand in single-threaded code into lambda functions that just happen to be run at the correct time. This preserves the simplicity of use I had with the multithreaded version, but without the need to have multiple threads.

Opinions?

**Review by kronbichler — APPROVED** [review]
This looks good to me for a simplification of interfaces in some application areas. I could run this through the step-37 benchmark (or similar ones) in case we apply this inside the DoFHandler, to see if we see any difference in timings.

**Comment by bangerth** [comment]
I've now moved these functions into `include/deal.II/base/mpi.h`. I think this patch is ready.

File: include/deal.II/base/mpi.h [hunk 1]
```diff
@@ -1996,6 +2232,144 @@ namespace Utilities
...
+      std::shared_ptr<std::vector<char>> send_buffer =
+        std::make_unique<std::vector<char>>(Utilities::pack(object, false));
+
+      // Now start the send, and store the result in a request object that
+      // we can then wait for later:
+      MPI_Request request;
+      const int   ierr =
+        MPI_Isend(send_buffer->data(),
+                  send_buffer->size(),
+                  mpi_type_id_for_type<decltype(*send_buffer->data())>,
```
Comment (masterleinad) [inline include/deal.II/base/mpi.h:2274]: This is always a `char`, right?
```suggestion
                  MPI_CHAR)>,
```

File: tests/mpi/isend_irecv_01.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Test the function compute_set_union() for set::vector and std::set.
```
Comment (masterleinad) [inline tests/mpi/isend_irecv_01.cc:17]: This comment doesn't seem to describe the test.

File: include/deal.II/base/mpi.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/base/mpi.h:2278]: Yes, but it is brittle to hard code it like this. A better approach is to have the MPI type deduced automatically so that it matches and references the object to be sent, like I do with `mpi_type_id_for_type`. I would like to leave it as is.

File: tests/mpi/isend_irecv_01.cc (hunk 2 above)
Comment (bangerth) [reply tests/mpi/isend_irecv_01.cc]: Yes, a copy-paste error :-)

File: include/deal.II/base/mpi.h [hunk 3]
```diff
@@ -1568,8 +1754,58 @@ namespace Utilities
...
+
+
+
+    template <typename T>
+    Future<T>::~Future()
+    {
+      // If there is a clean-up function, and if it has not been
+      // called yet, then do so. Note that we may not have a
+      // clean-up function (not even an empty one) if the current
+      // object has been moved from into another object.
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/base/mpi.h]: No, this was correct. What I want to say is that "this object has been moved from, into another object, and is now no longer usable". I'll fix this.

[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/base/mpi.h]: To be precise, this object was the right hand side of some move operation.

**Comment by bangerth** [comment]
Ping? The one failing tester is a Windoze machine that can't seem to upload its results. But everything compiled and tested ok.

**Comment by bangerth** [comment]
Ping?
